datalad/local/copy_file.py

```diff
@@ -0,0 +1,672 @@
+                    yield dict(
+                        path=str(src_path),
+                        status='impossible',
+                        message='recursion not enabled, omitting directory',
+                        **res_kwargs
+                    )
+                    continue
+
+                if dest_path and dest_path.name == '.git' or src_path.name == '.git':
+                    yield dict(
+                        path=str(src_path),
+                        status='impossible',
+                        message="refuse to place '.git' into destination dataset",
+                        **res_kwargs
+                    )
+                    continue
+
+                if not (dest_path or target_dir):
+                    yield dict(
+                        path=str(src_path),
+                        status='impossible',
+                        message='need destination path or target directory',
+                        **res_kwargs
+                    )
+                    continue
+
+                for src_file, dest_file in _yield_src_dest_filepaths(
+                        src_path, dest_path, target_dir=target_dir):
+                    if ds and ds.pathobj not in dest_file.parents:
+                        # take time to compose proper error
+                        dpath = str(target_dir if target_dir else dest_path)
+                        yield dict(
+                            path=dpath,
+                            status='error',
+                            message=(
+                                'reference dataset does not contain '
+                                'destination path: %s',
+                                dpath),
+                            **res_kwargs
+                        )
+                        # only recursion could yield further results, which would
+                        # all have the same issue, so call it over right here
+                        break
+                    for res in _copy_file(src_file, dest_file, cache=repo_cache):
+                        yield dict(
+                            res,
+                            **res_kwargs
+                        )
+                        if res.get('status', None) == 'ok':
+                            to_save.append(res['destination'])
+        finally:
+            # cleanup time
+            # TODO this could also be the place to stop lingering batch processes
+            done = set()
+            for _, repo_rec in repo_cache.items():
+                repo = repo_rec['repo']
+                if not repo or repo.pathobj in done:
+                    continue
+                tmp = repo_rec.get('tmp', None)
+                if tmp:
+                    try:
+                        tmp.rmdir()
+                    except OSError as e:
+                        lgr.warning(
+                            'Failed to clean up temporary directory: %s', e)
```

Comment:
Based on this [comment][0], I think @mih is intentionally avoiding it.  You guys might want to hash that out to avoid inconsistency and unnecessary churn in the code base.

[0]: https://github.com/datalad/datalad/pull/4203#discussion_r410191253

Reply:
thanks, left a comment there.  Here it is even higher level than debug -- a `warning` -- so I am ok to spend those few `ms` to possibly provide more diagnostic information if so desired.  I will also later today propose a PR to make it a "non-issue" and probably to demonstrate that it indeed was not any performance issue to start with.

Reply:
reread the code -- I was wrong in that I assumed that if no limit explicitly provided we would not include any traceback location information thus we could simply avoid it altogether if DATALAD_EXC_STR_TBLIMIT variable is not set.  But we are defaulting to limit of 1 (thus showing the file:lineno where exception was raised).   I think it is valuable even if in many cases we could match location where it was thrown by `git grep`ing the message (until it starts coming from python standard libraries etc).  So I will keep it as is and shortly push a docstring change (with no ci) to clarify the parameters etc

Reply:
did in did in 11932b3f4ae60fb4ee7798731a669d6e1598481e